plugins/system_upgrade.py

```diff
@@ -0,0 +1,713 @@
+
+
+# Translators: This string is only used in unit tests.
+_("the color of the sky")
+
+DOWNLOAD_FINISHED_ID = uuid.UUID('9348174c5cc74001a71ef26bd79d302e')
+REBOOT_REQUESTED_ID = uuid.UUID('fef1cc509d5047268b83a3a553f54b43')
+UPGRADE_STARTED_ID = uuid.UUID('3e0a5636d16b4ca4bbe5321d06c6aa62')
+UPGRADE_FINISHED_ID = uuid.UUID('8cec00a1566f4d3594f116450395f06c')
+
+ID_TO_IDENTIFY_BOOTS = UPGRADE_STARTED_ID
+
+PLYMOUTH = '/usr/bin/plymouth'
+
+RELEASEVER_MSG = _(
+    "Need a --releasever greater than the current system version.")
+DOWNLOAD_FINISHED_MSG = _(  # Translators: do not change "reboot" here
+    "Download complete! Use 'dnf {command} reboot' to start the upgrade.\n"
+    "To remove cached metadata and transaction use 'dnf {command} clean'")
+CANT_RESET_RELEASEVER = _(
+    "Sorry, you need to use 'download --releasever' instead of '--network'")
+
+STATE_VERSION = 2
+
+# --- Miscellaneous helper functions ------------------------------------------
+
+
+def reboot():
+    if os.getenv("DNF_SYSTEM_UPGRADE_NO_REBOOT", default=False):
+        logger.info(_("Reboot turned off, not rebooting."))
+    else:
+        Popen(["systemctl", "reboot"])
+
+
+def get_url_from_os_release():
+    key = "UPGRADE_GUIDE_URL="
+    for path in ["/etc/os-release", "/usr/lib/os-release"]:
+        try:
+            with open(path) as release_file:
+                for line in release_file:
+                    line = line.strip()
+                    if line.startswith(key):
+                        return line[len(key):].strip('"')
+        except IOError:
+            continue
+    return None
+
+
+# DNF-FIXME: dnf.util.clear_dir() doesn't delete regular files :/
+def clear_dir(path, ignore=[]):
+    if not os.path.isdir(path):
+        return
+
+    for entry in os.listdir(path):
+        fullpath = os.path.join(path, entry)
+        if fullpath in ignore:
+            continue
+        try:
+            if os.path.isdir(fullpath):
+                dnf.util.rm_rf(fullpath)
+            else:
+                os.unlink(fullpath)
+        except OSError:
+            pass
+
+
+def check_release_ver(conf, target=None):
+    if dnf.rpm.detect_releasever(conf.installroot) == conf.releasever:
+        raise CliError(RELEASEVER_MSG)
+    if target and target != conf.releasever:
+        # it's too late to set releasever here, so this can't work.
+        # (see https://bugzilla.redhat.com/show_bug.cgi?id=1212341)
+        raise CliError(CANT_RESET_RELEASEVER)
+
+
+def disable_blanking():
+    try:
+        tty = open('/dev/tty0', 'wb')
+        tty.write(b'\33[9;0]')
+    except Exception as e:
+        print(_("Screen blanking can't be disabled: %s") % e)
+
+
+def is_rhel_system_present():
+    ts = rpm.TransactionSet()
+    pkg = next(ts.dbMatch('provides', 'system-release'), None)
+    if not pkg:
+        logger.debug(_("No package providing 'system-release' was found."))
+        return False
+    return 'redhat-release' in pkg['provides']
```

Comment:
This check does not work as almost all Red Hat Linux derivatives would match on this. That includes Fedora, OpenMandriva, ROSA, and Mageia. You'll need to check via the `os-release` file instead. RHEL ships the `distro` Python module for this purpose.

Reply:
Thanks for the quick response. Why do you think so? 

AFAIK there should be a single installed package providing the `system-release` and this one provides the `*-release` specific for given distribution. F.e. on my Fedora 36 I have installed the `fedora-release-workstation` providing the `system-release` and this package provides also the `fedora-release` + also there is none package providing `redhat-release`.

Reply:
I guess it was removed from most `fedora-release-*` variants, but it exists in `mageia-release-Default`, `distro-release-OpenMandriva`, `rosa-release`, and others.

Reply:
Technically RHEL is a derivative of Fedora. Using redhat-release in any systems derived form Fedora is not technically correct and it is confusing. I also don't think that parsing any file on the system can help because each distribution can crate the same mess in file like with package provides. On the other hand it is only a warning.

Reply:
I realized that the `distro` module is already present in the `dnf-plugins-core`, therefore switching to this approach as there are no cons I can think of now.